libraries/types/pbft_block/src/pbft_block.cpp

```diff
@@ -11,22 +11,37 @@ namespace taraxa {
 PbftBlock::PbftBlock(bytes const& b) : PbftBlock(dev::RLP(b)) {}
 
 PbftBlock::PbftBlock(dev::RLP const& rlp) {
-  util::rlp_tuple(util::RLPDecoderRef(rlp, true), prev_block_hash_, dag_block_hash_as_pivot_, order_hash_,
-                  prev_state_root_hash_, period_, timestamp_, reward_votes_, signature_);
+  if (rlp.itemCount() == 9) {
```

Comment:
this will be problematic because I am adding another optional parameter into pbft block so possible itemCount will be 8,9 or 10. If size == 9, we won't be able to say which optional param is present.

if you declare extra_data as std::optional<bytes> and leave it empty, it will put 1 zero byre into the rlp, otherwise the whole extra data and we don't need to check itemCount. You can then use util::rlp_tuple for all params...

Reply:
This will make hard fork much more complex since we will need to have different format of pbft block before and after the hardfork and there will be a lot more places in the code where we have to check if we are dealing with an old format block or a new one. Because adding a zero byte in rlp changes the hash, so all the pre-fork blocks need to have the old format. I mean it should be the same case for your param as well. We will still need to check the item count because it could always be an old format block. I will think some more if there is some better way to handle this.

Reply:
ok give it a though. I don't see any other way how to do it at the moment...

Reply:
What do you think about putting the pillar block hash within the extra_data field so this way we only have one optional field? Ethereum standard for extra data is 32 bytes but we could expand it to more to fit both version and pillar block hash.

Reply:
it would work probably. Not sure if it is good or bad solution to put it there...

Reply:
I made some changes that the extra_data part as a separate optional structure within the pbft block. So you could add your change within there. So with your change pbft block rlp count would no longer change, only the rlp within the PbftBlockExtraData would change